File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
private:
    stack<char> brackets;
public:
    bool isValid(string s) {
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                brackets.push(s[i]);
            }
            else {
                if (brackets.empty()) {
                    return false;
                }
                if (s[i] == ')') {
                    if (brackets.top() != '(') {
                        return false;
                    }
                }
                if (s[i] == '}') {
                    if (brackets.top() != '{') {
                        return false;
                    }
                }
                if (s[i] == ']') {
                    if (brackets.top() != '[') {
                        return false;
                    }
                }
                brackets.pop();
            }
        }
        
        if (!brackets.empty()) {
            return false;
        }
        return true;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = Stack<Char>()

        for (i in s) {
            if (isOpeningBracket(i)) stack.push(i)
            else {
                if (stack.isEmpty()) return false
                if (isMatchingBracket(stack.peek(), i)) {
                    stack.pop()
                }
                else return false
            }
        }
        if (stack.isNotEmpty()) return false
        return true
    }

    private fun isOpeningBracket(bracket: Char): Boolean = bracket == '(' || bracket == '{' || bracket == '['

    private fun isClosingBracket(bracket: Char): Boolean = bracket == ')' || bracket == '}' || bracket == ']'

    private fun isMatchingBracket(first: Char, second: Char): Boolean {
        return when {
            first == '(' && second == ')' -> true
            first == '{' && second == '}' -> true
            first == '[' && second == ']' -> true
            else -> false
        }
    }
}
